In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_excel('/content/Online retail.xlsx',header=None)
df.head()

0
0  shrimp,almonds,avocado,vegetables mix,green gr...
1                             burgers,meatballs,eggs
2                                            chutney
3                                     turkey,avocado
4  mineral water,milk,energy bar,whole wheat rice...

as we see there are all items available in store are present on top then the bottom columns are the cusromers who purchased goods so now we have to print the combination of goods which are bought frequently using association rules

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


In [4]:
df.isnull().sum()

0    0
dtype: int64

as we see there are no null rows in the data which means each costumer putrchased at least 1 item from the store

In [5]:
df.duplicated().sum()

np.int64(2325)

as we see there are duplicates which effect the efficiency so we remove them

In [6]:
df.drop_duplicates(inplace=True)

In [7]:
from mlxtend.preprocessing import TransactionEncoder

# Convert each row's string into a list of items
transactions = df.iloc[:, 0].apply(lambda x: x.split(','))

# One-hot encode
te = TransactionEncoder()
te_array = te.fit(transactions).transform(transactions)

# Create basket format
basket = pd.DataFrame(te_array, columns=te.columns_)

# Convert boolean values to 0 and 1
basket = pd.DataFrame(te_array.astype(int), columns=te.columns_)

# Display the first few rows
print(basket.head())


    asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0           0        1                  1          0        1            0   
1           0        0                  0          0        0            0   
2           0        0                  0          0        0            0   
3           0        0                  0          0        1            0   
4           0        0                  0          0        0            0   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0      0               0          0            0  ...       0               1   
1      0               0          0            0  ...       0               0   
2      0               0          0            0  ...       0               0   
3      0               0          0            0  ...       1               0   
4      0               0          0            0  ...       0               0   

   water spray  white wine  whole weat flour

now the data is in good shape

# Association Rule Mining:

In [8]:
from mlxtend.frequent_patterns import apriori, association_rules

freq_itemsets = apriori(basket, min_support=0.05, use_colnames=True)
freq_itemsets[0:1000]

/usr/local/lib/python3.11/dist-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                            itemsets
0   0.113794                           (burgers)
1   0.103555                              (cake)
2   0.054869                         (champagne)
3   0.083849                           (chicken)
4   0.205178                         (chocolate)
5   0.060665                           (cookies)
6   0.071870                       (cooking oil)
7   0.208076                              (eggs)
8   0.083849                          (escalope)
9   0.192620                      (french fries)
10  0.059312                       (fresh bread)
11  0.084815                   (frozen smoothie)
12  0.129830                 (frozen vegetables)
13  0.070131                     (grated cheese)
14  0.169243                         (green tea)
15  0.135819                       (ground beef)
16  0.066461                     (herb & pepper)
17  0.065301                             (honey)
18  0.095054                    (low fat yogurt)
19  0.170015                              (milk)
20  0.299845                     (mineral water)
21  0.087713                         (olive oil)
22  0.125193                          (pancakes)
23  0.057573                            (salmon)
24  0.098918                            (shrimp)
25  0.070904                              (soup)
26  0.229521                         (spaghetti)
27  0.091963                          (tomatoes)
28  0.083269                            (turkey)
29  0.077859                  (whole wheat rice)
30  0.073223          (mineral water, chocolate)
31  0.055835              (spaghetti, chocolate)
32  0.070131               (mineral water, eggs)
33  0.051391                   (eggs, spaghetti)
34  0.050425  (mineral water, frozen vegetables)
35  0.058733        (mineral water, ground beef)
36  0.055835            (spaghetti, ground beef)
37  0.067813               (mineral water, milk)
38  0.050039                   (milk, spaghetti)
39  0.085008          (mineral water, spaghetti)

first let see the items which are bought frequently with support values

now we see pair of items frequently purchased by the customers

In [9]:
rules = association_rules(freq_itemsets, metric='lift', min_threshold=1)
rules

antecedents          consequents  antecedent support  \
0       (mineral water)          (chocolate)            0.299845   
1           (chocolate)      (mineral water)            0.205178   
2           (spaghetti)          (chocolate)            0.229521   
3           (chocolate)          (spaghetti)            0.205178   
4       (mineral water)               (eggs)            0.299845   
5                (eggs)      (mineral water)            0.208076   
6                (eggs)          (spaghetti)            0.208076   
7           (spaghetti)               (eggs)            0.229521   
8       (mineral water)  (frozen vegetables)            0.299845   
9   (frozen vegetables)      (mineral water)            0.129830   
10      (mineral water)        (ground beef)            0.299845   
11        (ground beef)      (mineral water)            0.135819   
12          (spaghetti)        (ground beef)            0.229521   
13        (ground beef)          (spaghetti)            0.135819   
14      (mineral water)               (milk)            0.299845   
15               (milk)      (mineral water)            0.170015   
16               (milk)          (spaghetti)            0.170015   
17          (spaghetti)               (milk)            0.229521   
18      (mineral water)          (spaghetti)            0.299845   
19          (spaghetti)      (mineral water)            0.229521   

    consequent support   support  confidence      lift  representativity  \
0             0.205178  0.073223    0.244201  1.190193               1.0   
1             0.299845  0.073223    0.356874  1.190193               1.0   
2             0.205178  0.055835    0.243266  1.185635               1.0   
3             0.229521  0.055835    0.272128  1.185635               1.0   
4             0.208076  0.070131    0.233892  1.124070               1.0   
5             0.299845  0.070131    0.337047  1.124070               1.0   
6             0.229521  0.051391    0.246982  1.076078               1.0   
7             0.208076  0.051391    0.223906  1.076078               1.0   
8             0.129830  0.050425    0.168170  1.295310               1.0   
9             0.299845  0.050425    0.388393  1.295310               1.0   
10            0.135819  0.058733    0.195876  1.442184               1.0   
11            0.299845  0.058733    0.432432  1.442184               1.0   
12            0.135819  0.055835    0.243266  1.791102               1.0   
13            0.229521  0.055835    0.411095  1.791102               1.0   
14            0.170015  0.067813    0.226160  1.330231               1.0   
15            0.299845  0.067813    0.398864  1.330231               1.0   
16            0.229521  0.050039    0.294318  1.282316               1.0   
17            0.170015  0.050039    0.218013  1.282316               1.0   
18            0.229521  0.085008    0.283505  1.235204               1.0   
19            0.299845  0.085008    0.370370  1.235204               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0   0.011701    1.051632       0.228235  0.169575   0.049097    0.300537  
1   0.011701    1.088674       0.201051  0.169575   0.081451    0.300537  
2   0.008742    1.050332       0.203212  0.147374   0.047920    0.257697  
3   0.008742    1.058537       0.196988  0.147374   0.055300    0.257697  
4   0.007741    1.033698       0.157645  0.160194   0.032599    0.285470  
5   0.007741    1.056115       0.139377  0.160194   0.053134    0.285470  
6   0.003633    1.023189       0.089275  0.133067   0.022663    0.235444  
7   0.003633    1.020397       0.091760  0.133067   0.019989    0.235444  
8   0.011496    1.046091       0.325620  0.132960   0.044060    0.278281  
9   0.011496    1.144778       0.262000  0.132960   0.126468    0.278281  
10  0.018008    1.074686       0.437914  0.155818   0.069496    0.314154  
11  0.018008    1.233606       0.354795  0.155818   0.189368    0.314154  
12  0.02466

- **Antecedents**: Items that appear together before a rule is applied.
- **Consequents**: Items that appear as a result of the antecedents.
- **Antecedent Support**: The proportion of transactions containing the antecedents.
- **Consequent Support**: The proportion of transactions containing the consequents.
- **Support**: The frequency of both antecedents and consequents appearing together.
- **Confidence**: The likelihood that the consequents appear when the antecedents are present.
- **Lift**: Measures how much more likely the consequents appear with antecedents than by chance.
- **Representativity**: Indicates how well the rule represents the dataset.
- **Leverage**: Shows the difference between observed and expected co-occurrence.
- **Conviction**: Measures how strongly the antecedents predict the consequents.
- **Zhang’s Metric**: A statistical measure of association strength.
- **Jaccard**: A similarity measure between antecedents and consequents.
- **Certainty**: Indicates the reliability of the rule.
- **Kulczynski**: A balance measure between confidence and support.

In [19]:
rules = association_rules(freq_itemsets, metric='conviction', min_threshold=1.2)
rules

antecedents      consequents  antecedent support  consequent support  \
0  (ground beef)  (mineral water)            0.135819            0.299845   
1  (ground beef)      (spaghetti)            0.135819            0.229521   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.058733    0.432432  1.442184               1.0  0.018008    1.233606   
1  0.055835    0.411095  1.791102               1.0  0.024661    1.308326   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.354795  0.155818   0.189368    0.314154  
1       0.511102  0.180400   0.235664    0.327181

# Analysis and Interpretation:

### **Detailed Analysis of Key Rules**
1. **Mineral Water → Chocolate**
   - **Confidence:** 24.42% (Customers who buy mineral water have a 24.42% chance of also buying chocolate.)
   - **Lift:** 1.19 (Chocolate is 19% more likely to be bought with mineral water than by chance.)
   - **Insight:** This suggests that customers who purchase mineral water may also be looking for snacks or indulgent items.

2. **Spaghetti → Ground Beef**
   - **Confidence:** 24.33% (Customers who buy spaghetti have a 24.33% chance of also buying ground beef.)
   - **Lift:** 1.79 (Ground beef is 79% more likely to be bought with spaghetti than by chance.)
   - **Insight:** This strong association indicates meal preparation habits—customers buying spaghetti are likely planning to cook a pasta dish that includes ground beef.

3. **Frozen Vegetables → Mineral Water**
   - **Confidence:** 38.84% (Customers who buy frozen vegetables have a 38.84% chance of also buying mineral water.)
   - **Lift:** 1.29 (Mineral water is 29% more likely to be bought with frozen vegetables than by chance.)
   - **Insight:** This suggests a health-conscious customer segment that prefers frozen vegetables and mineral water, possibly indicating a preference for nutritious meal choices.

### **Business Implications**
- **Product Placement:** Items frequently bought together should be placed near each other in stores to encourage impulse purchases.
- **Bundling & Promotions:** Discounts on bundles like "spaghetti + ground beef" or "mineral water + chocolate" could boost sales.
- **Inventory Optimization:** Stocking patterns can be adjusted based on high-confidence associations to meet demand efficiently.
- **Targeted Marketing:** Personalized recommendations can be made based on these associations, such as offering discounts on chocolate to customers who frequently buy mineral water.

Would you like me to explore specific strategies for leveraging these insights further?



# Interview Questions:

### 1. **Lift and Its Importance in Association Rules**
Lift measures how much more likely the consequent (Y) appears with the antecedent (X) than it would by chance. It is calculated as:

\[
\text{Lift} = \frac{\text{Confidence}(X \Rightarrow Y)}{\text{Support}(Y)}
\]

- **Lift > 1**: Indicates a strong positive association between X and Y.
- **Lift = 1**: Suggests no association (X and Y appear together as expected).
- **Lift < 1**: Implies a negative association (X and Y appear together less than expected).

Lift is crucial because it helps determine whether a rule is meaningful beyond simple co-occurrence.

### 2. **Support and Confidence: Meaning & Calculation**
- **Support**: Measures how frequently an itemset appears in the dataset.
  
  \[
  \text{Support}(X \Rightarrow Y) = \frac{\text{Transactions containing } X \text{ and } Y}{\text{Total transactions}}
  \]

- **Confidence**: Indicates the likelihood that Y appears when X is present.
  
  \[
  \text{Confidence}(X \Rightarrow Y) = \frac{\text{Support}(X \Rightarrow Y)}{\text{Support}(X)}
  \]

These metrics help assess the reliability of association rules.

### 3. **Limitations & Challenges of Association Rule Mining**
- **High Computational Cost**: Large datasets require significant processing power.
- **Too Many Rules**: Can generate excessive rules, making interpretation difficult.
- **Low Support for Rare Items**: Rare but valuable associations may be overlooked.
- **Overfitting**: Some rules may not generalize well beyond the dataset.
- **Lack of Context**: Rules show correlations but don’t explain causation.
